service/quote-server/tdx_lib/src/tdx_lib.rs

```diff
@@ -0,0 +1,387 @@
+use std::mem;
+use std::os::unix::io::AsRawFd;
+use std::path::Path;
+use std::ptr;
+use std::result::Result;
+use std::result::Result::Ok;
+
+#[repr(C)]
+pub struct tdx10_report_req {
+    subtype: u8,
+    reportdata: u64,
+    rpd_len: u32,
+    tdreport: u64,
+    tdr_len: u32,
+}
+
+#[repr(C)]
+pub struct tdx15_report_req {
+    reportdata: [u8; REPORT_DATA_LEN as usize],
+    tdreport: [u8; TDX_REPORT_LEN as usize],
+}
+
+#[repr(C)]
+pub struct qgs_msg_header {
+    major_version: u16,
+    minor_version: u16,
+    msg_type: u32,
+    size: u32,       // size of the whole message, include this header, in byte
+    error_code: u32, // used in response only
+}
+
+#[repr(C)]
+pub struct qgs_msg_get_quote_req {
+    header: qgs_msg_header,                        // header.type = GET_QUOTE_REQ
+    report_size: u32,                              // cannot be 0
+    id_list_size: u32,                             // length of id_list, in byte, can be 0
+    report_id_list: [u8; TDX_REPORT_LEN as usize], // report followed by id list
+}
+
+#[repr(C)]
+pub struct tdx_quote_hdr {
+    version: u64,                       // Quote version, filled by TD
+    status: u64,                        // Status code of Quote request, filled by VMM
+    in_len: u32,                        // Length of TDREPORT, filled by TD
+    out_len: u32,                       // Length of Quote, filled by VMM
+    data_len_be_bytes: [u8; 4],         // big-endian 4 bytes indicate the size of data following
+    data: [u8; TDX_QUOTE_LEN as usize], // Actual Quote data or TDREPORT on input
+}
+
+#[repr(C)]
+pub struct tdx_quote_req {
+    buf: u64,
+    len: u64,
+}
+
+#[repr(C)]
+pub struct qgs_msg_get_quote_resp {
+    header: qgs_msg_header,        // header.type = GET_QUOTE_RESP
+    selected_id_size: u32,         // can be 0 in case only one id is sent in request
+    quote_size: u32,               // length of quote_data, in byte
+    id_quote: [u8; TDX_QUOTE_LEN], // selected id followed by quote
+}
+
+pub enum TdxType {
+    TDX10,
+    TDX15,
+}
+
+const REPORT_DATA_LEN: u32 = 64;
+const TDX_REPORT_LEN: u32 = 1024;
+const TDX_QUOTE_LEN: usize = 4 * 4096;
+
+pub struct TdxInfo {
+    tdx_version: TdxType,
+    device_node: File,
+}
+
+impl TdxInfo {
+    fn new(_tdx_version: TdxType, _device_node: File) -> Self {
+        TdxInfo {
+            tdx_version: _tdx_version,
+            device_node: _device_node,
+        }
+    }
+}
+
+fn get_tdx_version() -> TdxType {
+    if Path::new("/dev/tdx-guest").exists() {
+        TdxType::TDX10
+    } else if Path::new("/dev/tdx_guest").exists() {
+        TdxType::TDX15
+    } else if Path::new("/dev/tdx-attest").exists() {
+        panic!("get_tdx_version: Deprecated device node /dev/tdx-attest, please upgrade to use /dev/tdx-guest or /dev/tdx_guest");
+    } else {
+        panic!("get_tdx_version: no TDX device found!");
+    }
+}
+
+pub fn get_tdx_report(report_data: String) -> Result<Vec<u8>, anyhow::Error> {
```

Comment:
Using `TD report` instead of `TDX report`, please follow the naming in the spec.

Reply:
updated!